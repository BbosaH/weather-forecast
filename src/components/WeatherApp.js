import React, { Component } from "react";
import "babel-polyfill";
import "../styles/styles.css";
import WeatherNavHeader from "./WeatherNavHeader";
import WeatherAdvice from "./WeatherAdvice";
import WeatherCardContainer from "./WeatherCardContainer";
import WeatherFooter from "./WeatherFooter";
import API, { API_AUTH_ID, ZIP_REGEX } from "../api/api";
import {
  formatDate,
  resetObject,
  processWeatherData,
  getWeatherData
} from "../utility/utility";
/**
 * Container component to manage state of Other components
 * Contains functions to retrieve and process data from Api
 * Does not take in any props.
 */
class WeatherApp extends Component {
  constructor(props) {
    super(props);
    this.state = {
      cityName: "Brooklyn",
      zipCode: "",
      returned_data: false,
      weather_data: {
        city: {
          id: "",
          name: "",
          country: ""
        },
        forecast_weathers: [
          {
            dt: "",
            today: false,
            date: "",
            day: "",
            low_temp: "",
            high_temp: "",
            weather_description: "",
            weather_icon: ""
          }
        ]
      },
      dayarray: [
        "Sunday",
        "Monday",
        "Tuesday",
        "Wednesday",
        "Thursday",
        "Friday",
        "Saturday"
      ],
      iconObj: {
        snow: "wi-day-snow",
        rain: "wi-day-rain"
      }
    };
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSearchClick = this.handleSearchClick.bind(this);
  }

  /**
   * event handler method invoked on change of any value in search input for weather forecast
   *  */

  handleInputChange(event) {
    let val = event.target.value;
    this.setState(prevState => {
      return ZIP_REGEX.test(val)
        ? {
            zipCode: val,
            cityName: val
          }
        : {
            cityName: val,
            zipCode: ""
          };
    });
  }

  componentDidMount() {
    this.handleSearchClick();
  }
  /**
   * Method executed on click of search button to retrieve new data
   * Also called within ComponentDidMount method on start for default city data.
   */
  handleSearchClick() {
    let todayDateStr = formatDate(new Date()).toString();
    let dataObject = getWeatherData(
      API,
      this.state.zipCode,
      this.state.cityName,
      API_AUTH_ID
    );
    dataObject.then(result => {
      let processed_data = processWeatherData(
        result,
        todayDateStr,
        this.state.dayarray,
        resetObject()
      );
      if (!processed_data) {
        this.setState(prevState => {
          return {
            returned_data: false
          };
        });
      } else {
        this.setState(prevState => {
          return processed_data;
        });
      }
    });
  }

  render() {
    return (
      <React.Fragment>
        <div className="background" />
        <WeatherNavHeader />
        <WeatherAdvice
          cityName={this.state.cityName}
          handleChange={this.handleInputChange}
          handleClick={this.handleSearchClick}
          weatherToday={this.state.weather_data.forecast_weathers[0]}
          weatherCity={this.state.weather_data.city}
        />
        <WeatherCardContainer
          returnedData={this.state.returned_data}
          weatherData={this.state.weather_data}
        />
        <WeatherFooter />
      </React.Fragment>
    );
  }
}

export default WeatherApp;
