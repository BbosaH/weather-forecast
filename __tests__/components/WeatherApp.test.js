import React from "react";
import { shallow } from "enzyme";
import WeatherApp from "../../src/components/WeatherApp";

describe("<App />", () => {
  describe("render()", () => {
    test("renders the component", () => {
      const wrapper = shallow(
        <WeatherApp
          dayarray={[
            "Sunday",
            "Monday",
            "Tuesday",
            "Wednesday",
            "Thursday",
            "Friday",
            "Saturday"
          ]}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
});
