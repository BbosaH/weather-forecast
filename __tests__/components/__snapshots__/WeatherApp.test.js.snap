// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> render() renders the component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WeatherApp
    dayarray={
      Array [
        "Sunday",
        "Monday",
        "Tuesday",
        "Wednesday",
        "Thursday",
        "Friday",
        "Saturday",
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <div
          className="background"
        />,
        <WeatherNavHeader />,
        <WeatherAdvice
          cityName="Brooklyn"
          handleChange={[Function]}
          handleClick={[Function]}
          weatherCity={
            Object {
              "country": "",
              "id": "",
              "name": "",
            }
          }
          weatherToday={
            Object {
              "date": "",
              "dt": "",
              "high_temp": "",
              "low_temp": "",
              "myDay": "",
              "today": false,
              "weather_description": "",
              "weather_icon": "",
            }
          }
        />,
        <WeatherCardContainer
          returnedData={false}
          weatherData={
            Object {
              "city": Object {
                "country": "",
                "id": "",
                "name": "",
              },
              "forecast_weathers": Array [
                Object {
                  "date": "",
                  "dt": "",
                  "high_temp": "",
                  "low_temp": "",
                  "myDay": "",
                  "today": false,
                  "weather_description": "",
                  "weather_icon": "",
                },
              ],
            }
          }
        />,
        <WeatherFooter />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "background",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "cityName": "Brooklyn",
          "handleChange": [Function],
          "handleClick": [Function],
          "weatherCity": Object {
            "country": "",
            "id": "",
            "name": "",
          },
          "weatherToday": Object {
            "date": "",
            "dt": "",
            "high_temp": "",
            "low_temp": "",
            "myDay": "",
            "today": false,
            "weather_description": "",
            "weather_icon": "",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "returnedData": false,
          "weatherData": Object {
            "city": Object {
              "country": "",
              "id": "",
              "name": "",
            },
            "forecast_weathers": Array [
              Object {
                "date": "",
                "dt": "",
                "high_temp": "",
                "low_temp": "",
                "myDay": "",
                "today": false,
                "weather_description": "",
                "weather_icon": "",
              },
            ],
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <div
            className="background"
          />,
          <WeatherNavHeader />,
          <WeatherAdvice
            cityName="Brooklyn"
            handleChange={[Function]}
            handleClick={[Function]}
            weatherCity={
              Object {
                "country": "",
                "id": "",
                "name": "",
              }
            }
            weatherToday={
              Object {
                "date": "",
                "dt": "",
                "high_temp": "",
                "low_temp": "",
                "myDay": "",
                "today": false,
                "weather_description": "",
                "weather_icon": "",
              }
            }
          />,
          <WeatherCardContainer
            returnedData={false}
            weatherData={
              Object {
                "city": Object {
                  "country": "",
                  "id": "",
                  "name": "",
                },
                "forecast_weathers": Array [
                  Object {
                    "date": "",
                    "dt": "",
                    "high_temp": "",
                    "low_temp": "",
                    "myDay": "",
                    "today": false,
                    "weather_description": "",
                    "weather_icon": "",
                  },
                ],
              }
            }
          />,
          <WeatherFooter />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "background",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "cityName": "Brooklyn",
            "handleChange": [Function],
            "handleClick": [Function],
            "weatherCity": Object {
              "country": "",
              "id": "",
              "name": "",
            },
            "weatherToday": Object {
              "date": "",
              "dt": "",
              "high_temp": "",
              "low_temp": "",
              "myDay": "",
              "today": false,
              "weather_description": "",
              "weather_icon": "",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "returnedData": false,
            "weatherData": Object {
              "city": Object {
                "country": "",
                "id": "",
                "name": "",
              },
              "forecast_weathers": Array [
                Object {
                  "date": "",
                  "dt": "",
                  "high_temp": "",
                  "low_temp": "",
                  "myDay": "",
                  "today": false,
                  "weather_description": "",
                  "weather_icon": "",
                },
              ],
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
